### ISSUE TITLE
improve documentation for common ctx shortcuts

### ISSUE THREAD

**Issue by dgoldstein0** [bazelbuild/bazel#6908]
### Description of the problem / feature request:

ctx has a lot of little magical features such as `ctx.runfiles(collect_default=True)` or `ctx.files.<foo>`.  While these features are nice for convenience, the docs on them are rather thin - they often say what they do, without spelling out what they are actually short for.  This can make it hard to understand:
- what the performance of these things is like - e.g. ctx.files expands depsets, but you can't tell that from it's docs (#6496)
- what to replace these things with when they stop being the right idiom to use; e.g. `ctx.runfiles(collect_default=True)` can be wrong thing to use for rules that want to compile files in their `srcs`, as it'll include the inputs in the runfiles which may not be desired. (example issue: transpiling .js to .js, if you include srcs in runfiles then you end up with the srcs and compiled outputs generating conflicting symlinks)

The docs should be updated to include "<short form> is equivalent to <more verbose thing>".  off the top of my head the following could use this:
- ctx.file
- ctx.files
- ctx.runfiles(collect_default=True) & ctx.runfiles(collect_data= True)

### Feature requests: what underlying problem are you trying to solve with this feature?

Improve documentation of ctx members (https://docs.bazel.build/versions/master/skylark/lib/ctx.html)

**Comment by Globegitter** [comment]
Yeah I think this will also go a long way with helping to understand the `DefaultInfo` provider better and being able to sett the runfiles correctly.

Also see https://github.com/bazelbuild/bazel/issues/3015

**Comment by dgoldstein0** [comment]
I still believe this would be valuable.  I tried to remove some of our usage of some of these magical, no longer recommended things, and got some nasty surprises because I didn't know what equivalents I should replace them with and guessed wrong.  So a good chunk of the value here is useful to get the community off these apis.

**Comment by dgoldstein0** [comment]
This is still relevant.

The docs for ctx.files do mention an "equivalent" construct, but I think it's wrong - it's written like DefaultInfo.files is a list, when I think it's usually a depset.

ctx.runfiles(collect_default=True) is only documented to say that it shouldn't be used, which doesn't actually tell us what it does to know how to migrate off it.
